File: Src/Visual Studio/SDK/C#/Eruru.ChatRobotRPC For Shared Project/ChatRobotMessage.cs
using System;

namespace Eruru.ChatRobotRPC {

	/// <summary>
	/// 聊天机器人消息
	/// </summary>
	public class ChatRobotMessage {

		/// <summary>
		/// 所属实例
		/// </summary>
		public ChatRobot ChatRobot { get; set; }
		/// <summary>
		/// 来源
		/// </summary>
		public ChatRobotMessageType Type { get; set; }
		/// <summary>
		/// 机器人
		/// </summary>
		public long Robot { get; set; }
		/// <summary>
		/// 发送者
		/// </summary>
		public long QQ { get; set; }
		/// <summary>
		/// 来源群
		/// </summary>
		public long Group { get; set; }
		/// <summary>
		/// 内容
		/// </summary>
		public string Text { get; set; }
		/// <summary>
		/// 与ID配合用于防撤回
		/// </summary>
		public long Number { get; set; }
		/// <summary>
		/// 与Number配合用于防撤回
		/// </summary>
		public long ID { get; set; }
		/// <summary>
		/// 接收时间
		/// </summary>
		public DateTime ReceivedTime { get; set; }

		/// <summary>
		/// 完整构造聊天机器人消息
		/// </summary>
		/// <param name="chatRobot">调用实例</param>
		/// <param name="type">消息类型</param>
		/// <param name="robot">响应机器人</param>
		/// <param name="group">来源群</param>
		/// <param name="qq">来源QQ</param>
		/// <param name="text">内容</param>
		/// <param name="number">消息序号</param>
		/// <param name="id">消息ID</param>
		/// <param name="receivedTime">接收消息时间</param>
		public ChatRobotMessage (ChatRobot chatRobot, ChatRobotMessageType type, long robot, long group, long qq, string text, long number, long id,
			DateTime receivedTime
		) {
			ChatRobot = chatRobot;
			Type = type;
			Robot = robot;
			QQ = qq;
			Group = group;
			Text = text;
			Number = number;
			ID = id;
			ReceivedTime = receivedTime;
		}
		/// <summary>
		/// 构造聊天机器人消息，无需指定接收时间，默认为DateTime.Now
		/// </summary>
		/// <param name="chatRobot">调用实例</param>
		/// <param name="type">消息类型</param>
		/// <param name="robot">响应机器人</param>
		/// <param name="group">来源群</param>
		/// <param name="qq">来源QQ</param>
		/// <param name="text">内容</param>
		/// <param name="number">消息序号</param>
		/// <param name="id">消息ID</param>
		public ChatRobotMessage (ChatRobot chatRobot, ChatRobotMessageType type, long robot, long group, long qq, string text, long number, long id)
			: this (chatRobot, type, robot, group, qq, text, number, id, DateTime.Now
		) {

		}

		/// <summary>
		/// 回复消息
		/// </summary>
		/// <param name="message">内容</param>
		/// 	/// <param name="isAnonymous">是否匿名（仅Pro有效）</param>
		public void Reply (string message, bool isAnonymous = false) {
			Reply (message, ChatRobotSendMessageType.Text, isAnonymous);
		}
		/// <summary>
		/// 回复消息
		/// </summary>
		/// <param name="message">内容</param>
		/// 	/// <param name="isAnonymous">是否匿名（仅Pro有效）</param>
		public void Reply (object message, bool isAnonymous = false) {
			Reply (message?.ToString (), isAnonymous);
		}

		/// <summary>
		/// 回复Json消息
		/// </summary>
		/// <param name="message">内容</param>
		/// 	/// <param name="isAnonymous">是否匿名（仅Pro有效）</param>
		public void ReplyJson (string message, bool isAnonymous = false) {
			Reply (message, ChatRobotSendMessageType.Json, isAnonymous);
		}
		/// <summary>
		/// 回复Json消息
		/// </summary>
		/// <param name="message">内容</param>
		/// 	/// <param name="isAnonymous">是否匿名（仅Pro有效）</param>
		public void ReplyJson (object message, bool isAnonymous = false) {
			ReplyJson (message?.ToString (), isAnonymous);
		}

		/// <summary>
		/// 回复Xml消息
		/// </summary>
		/// <param name="message">内容</param>
		/// <param name="isAnonymous">是否匿名（仅Pro有效）</param>
		public void ReplyXml (string message, bool isAnonymous = false) {
			Reply (message, ChatRobotSendMessageType.Xml, isAnonymous);
		}
		/// <summary>
		/// 回复Xml消息
		/// </summary>
		/// <param name="message">群内</param>
		/// <param name="isAnonymous">是否匿名（仅Pro有效）</param>
		public void ReplyXml (object message, bool isAnonymous = false) {
			ReplyXml (message?.ToString (), isAnonymous);
		}

		/// <summary>
		/// 发送消息
		/// </summary>
		/// <param name="chatRobot">调用实例</param>
		/// <param name="type">目标类型</param>
		/// <param name="robot">机器人</param>
		/// <param name="message">内容</param>
		/// <param name="group">目标群</param>
		/// <param name="qq">目标QQ</param>
		/// <param name="isAnonymous">是否匿名（仅Pro有效）</param>
		/// <param name="sendType">文本、Json、Xml消息</param>
		public static void Send (ChatRobot chatRobot, ChatRobotMessageType type, long robot, string message, long group = default, long qq = default,
			bool isAnonymous = false, ChatRobotSendMessageType sendType = ChatRobotSendMessageType.Text
		) {
			switch (type) {
				case ChatRobotMessageType.Friend:
					switch (sendType) {
						case ChatRobotSendMessageType.Text:
							chatRobot.SendFriendMessage (robot, qq, message);
							break;
						case ChatRobotSendMessageType.Json:
							chatRobot.SendFriendJsonMessage (robot, qq, message);
							break;
						case ChatRobotSendMessageType.Xml:
							chatRobot.SendFriendXmlMessage (robot, qq, message);
							break;
						default:
							throw new NotImplementedException ($"{type}.{sendType}");
					}
					break;
				case ChatRobotMessageType.GroupTemp:
					switch (sendType) {
						case ChatRobotSendMessageType.Text:
							chatRobot.SendGroupTempMessage (robot, group, qq, message);
							break;
						case ChatRobotSendMessageType.Json:
							chatRobot.SendGroupTempJsonMessage (robot, group, qq, message);
							break;
						case ChatRobotSendMessageType.Xml:
							chatRobot.SendGroupTempXmlMessage (robot, group, qq, message);
							break;
						default:
							throw new NotImplementedException ($"{type}.{sendType}");
					}
					break;
				case ChatRobotMessageType.DiscussTemp:
					switch (sendType) {
						case ChatRobotSendMessageType.Text:
							chatRobot.SendDiscussTempMessage (robot, group, qq, message);
							break;
						case ChatRobotSendMessageType.Json:
							chatRobot.SendDiscussTempJsonMessage (robot, group, qq, message);
							break;
						case ChatRobotSendMessageType.Xml:
							chatRobot.SendDiscussTempXmlMessage (robot, group, qq, message);
							break;
						default:
							throw new NotImplementedException ($"{type}.{sendType}");
					}
					break;
				case ChatRobotMessageType.WebpageTemp:
					switch (sendType) {
						case ChatRobotSendMessageType.Text:
							chatRobot.SendWebpageTempMessage (robot, qq, message);
							break;
						case ChatRobotSendMessageType.Json:
							chatRobot.SendWebpageTempJsonMessage (robot, qq, message);
							break;
						case ChatRobotSendMessageType.Xml:
							chatRobot.SendWebpageTempXmlMessage (robot, qq, message);
							break;
						default:
							throw new NotImplementedException ($"{type}.{sendType}");
					}
					break;
				case ChatRobotMessageType.FriendVerificationReply:
					switch (sendType) {
						case ChatRobotSendMessageType.Text:
							chatRobot.SendFriendVerificationReplyMessage (robot, qq, message);
							break;
						case ChatRobotSendMessageType.Json:
							chatRobot.SendFriendVerificationReplyJsonMessage (robot, qq, message);
							break;
						case ChatRobotSendMessageType.Xml:
							chatRobot.SendFriendVerificationReplyXmlMessage (robot, qq, message);
							break;
						default:
							throw new NotImplementedException ($"{type}.{sendType}");
					}
					break;
				case ChatRobotMessageType.Group:
					switch (sendType) {
						case ChatRobotSendMessageType.Text:
							chatRobot.SendGroupMessage (robot, group, message, isAnonymous);
							break;
						case ChatRobotSendMessageType.Json:
							chatRobot.SendGroupJsonMessage (robot, group, message, isAnonymous);
							break;
						case ChatRobotSendMessageType.Xml:
							chatRobot.SendGroupXmlMessage (robot, group, message, isAnonymous);
							break;
						default:
							throw new NotImplementedException ($"{type}.{sendType}");
					}
					break;
				case ChatRobotMessageType.Discuss:
					switch (sendType) {
						case ChatRobotSendMessageType.Text:
							chatRobot.SendDiscussMessage (robot, group, message);
							break;
						case ChatRobotSendMessageType.Json:
							chatRobot.SendDiscussJsonMessage (robot, group, message);
							break;
						case ChatRobotSendMessageType.Xml:
							chatRobot.SendDiscussXmlMessage (robot, group, message);
							break;
						default:
							throw new NotImplementedException ($"{type}.{sendType}");
					}
					break;
				default:
					throw new NotImplementedException (type.ToString ());
			}
		}
		/// <summary>
		/// 发送消息
		/// </summary>
		/// <param name="chatRobot">调用实例</param>
		/// <param name="type">目标类型</param>
		/// <param name="robot">机器人</param>
		/// <param name="message">内容</param>
		/// <param name="group">目标群</param>
		/// <param name="qq">目标QQ</param>
		/// <param name="isAnonymous">是否匿名（仅Pro有效）</param>
		/// <param name="sendType">文本、Json、Xml消息</param>
		public static void Send (ChatRobot chatRobot, ChatRobotMessageType type, long robot, object message, long group = default, long qq = default,
			bool isAnonymous = false,
			ChatRobotSendMessageType sendType = ChatRobotSendMessageType.Text
		) {
			Send (chatRobot, type, robot, message?.ToString (), group, qq, isAnonymous, sendType);
		}

		/// <summary>
		/// 返回消息内容（ChatRobotMessage.Text）
		/// </summary>
		/// <returns></returns>
		public override string ToString () {
			return Text;
		}

		void Reply (string message, ChatRobotSendMessageType type, bool isAnonymous) {
			Send (ChatRobot, Type, Robot, message, Group, QQ, isAnonymous, type);
		}

	}

}

File: Src/Intelij IDEA/SDK/Java/Eruru.ChatRobotRPC/src/main/java/org/eruru/chatrobotrpc/ChatRobotMessage.java
package org.eruru.chatrobotrpc;

import javafx.util.Pair;
import org.eruru.chatrobotrpc.enums.ChatRobotMessageType;
import org.eruru.chatrobotrpc.enums.ChatRobotSendMessageType;

import java.util.Date;
import java.util.List;

/**
 * 聊天机器人消息
 */
public class ChatRobotMessage {

	private ChatRobot chatRobot;
	private ChatRobotMessageType type;
	private long robot;
	private long qq;
	private long group;
	private String text;
	private long number;
	private long id;
	private Date receivedTime;

	/**
	 * 所属实例
	 */
	public ChatRobot getChatRobot () {
		return chatRobot;
	}

	/**
	 * 所属实例
	 */
	public void setChatRobot (ChatRobot chatRobot) {
		this.chatRobot = chatRobot;
	}

	/**
	 * 来源
	 */
	public ChatRobotMessageType getType () {
		return type;
	}

	/**
	 * 来源
	 */
	public void setType (ChatRobotMessageType type) {
		this.type = type;
	}

	/**
	 * 机器人
	 */
	public long getRobot () {
		return robot;
	}

	/**
	 * 机器人
	 */
	public void setRobot (long robot) {
		this.robot = robot;
	}

	/**
	 * 发送者
	 */
	public long getQQ () {
		return qq;
	}

	/**
	 * 发送者
	 */
	public void setQQ (long qq) {
		this.qq = qq;
	}

	/**
	 * 来源群
	 */
	public long getGroup () {
		return group;
	}

	/**
	 * 来源群
	 */
	public void setGroup (long group) {
		this.group = group;
	}

	/**
	 * 内容
	 */
	public String getText () {
		return text;
	}

	/**
	 * 内容
	 */
	public void setText (String text) {
		this.text = text;
	}

	/**
	 * 与ID配合用于防撤回
	 */
	public long getNumber () {
		return number;
	}

	/**
	 * 与ID配合用于防撤回
	 */
	public void setNumber (long number) {
		this.number = number;
	}

	/**
	 * 与Number配合用于防撤回
	 */
	public long getID () {
		return id;
	}

	/**
	 * 与Number配合用于防撤回
	 */
	public void setID (long id) {
		this.id = id;
	}

	/**
	 * 接收时间
	 */
	public Date getReceivedTime () {
		return receivedTime;
	}

	/**
	 * 接收时间
	 */
	public void setReceivedTime (Date receivedTime) {
		this.receivedTime = receivedTime;
	}

	/**
	 * 完整构造聊天机器人消息
	 *
	 * @param chatRobot    调用实例
	 * @param type         消息类型
	 * @param robot        响应机器人
	 * @param group        来源群
	 * @param qq           来源QQ
	 * @param text         内容
	 * @param number       消息序号
	 * @param id           消息ID
	 * @param receivedTime 接收消息时间
	 */
	public ChatRobotMessage (ChatRobot chatRobot, ChatRobotMessageType type, long robot, long group, long qq, String text, long number, long id, Date receivedTime) {
		this.chatRobot = chatRobot;
		this.type = type;
		this.robot = robot;
		this.qq = qq;
		this.group = group;
		this.text = text;
		this.number = number;
		this.id = id;
		this.receivedTime = receivedTime;
	}

	/**
	 * 构造聊天机器人消息，无需指定接收时间，默认为new Date ()
	 *
	 * @param chatRobot 调用实例
	 * @param type      消息类型
	 * @param robot     响应机器人
	 * @param group     来源群
	 * @param qq        来源QQ
	 * @param text      内容
	 * @param number    消息序号
	 * @param id        消息ID
	 */
	public ChatRobotMessage (ChatRobot chatRobot, ChatRobotMessageType type, long robot, long group, long qq, String text, long number, long id) {
		this (chatRobot, type, robot, group, qq, text, number, id, new Date ());
	}

	/**
	 * 回复消息
	 *
	 * @param message     内容
	 * @param isAnonymous 是否匿名（仅Pro有效）
	 */
	public void reply (String message, boolean isAnonymous) {
		reply (message, ChatRobotSendMessageType.text, isAnonymous);
	}

	/**
	 * 回复消息
	 *
	 * @param message 内容
	 */
	public void reply (String message) {
		reply (message, ChatRobotSendMessageType.text, false);
	}

	/**
	 * 回复消息
	 *
	 * @param message     内容
	 * @param isAnonymous 是否匿名（仅Pro有效）
	 */
	public void reply (Object message, boolean isAnonymous) {
		reply (message, ChatRobotSendMessageType.text, isAnonymous);
	}

	/**
	 * 回复消息
	 *
	 * @param message 内容
	 */
	public void reply (Object message) {
		reply (message, ChatRobotSendMessageType.text, false);
	}

	/**
	 * 回复消息
	 *
	 * @param format      格式化字符串
	 * @param isAnonymous 是否匿名（仅Pro有效）
	 * @param args        可变长参数
	 */
	public void reply (String format, boolean isAnonymous, Object... args) {
		reply (format, ChatRobotSendMessageType.text, isAnonymous, args);
	}

	/**
	 * 回复消息
	 *
	 * @param format 格式化字符串
	 * @param args   可变长参数
	 */
	public void reply (String format, Object... args) {
		reply (format, ChatRobotSendMessageType.text, false, args);
	}

	/**
	 * 回复Json消息
	 *
	 * @param message     内容
	 * @param isAnonymous 是否匿名（仅Pro有效）
	 */
	public void replyJson (String message, boolean isAnonymous) {
		reply (message, ChatRobotSendMessageType.json, isAnonymous);
	}

	/**
	 * 回复Json消息
	 *
	 * @param message 内容
	 */
	public void replyJson (String message) {
		reply (message, ChatRobotSendMessageType.json, false);
	}

	/**
	 * 回复Json消息
	 *
	 * @param message     内容
	 * @param isAnonymous 是否匿名（仅Pro有效）
	 */
	public void replyJson (Object message, boolean isAnonymous) {
		reply (message, ChatRobotSendMessageType.json, isAnonymous);
	}

	/**
	 * 回复Json消息
	 *
	 * @param message 内容
	 */
	public void replyJson (Object message) {
		reply (message, ChatRobotSendMessageType.json, false);
	}

	/**
	 * 回复Json消息
	 *
	 * @param format      格式化字符串
	 * @param isAnonymous 是否匿名（仅Pro有效）
	 * @param args        可变长参数
	 */
	public void replyJson (String format, boolean isAnonymous, Object... args) {
		reply (format, ChatRobotSendMessageType.json, isAnonymous, args);
	}

	/**
	 * 回复Json消息
	 *
	 * @param format 格式化字符串
	 * @param args   可变长参数
	 */
	public void replyJson (String format, Object... args) {
		reply (format, ChatRobotSendMessageType.json, false, args);
	}

	/**
	 * 回复Xml消息
	 *
	 * @param message     内容
	 * @param isAnonymous 是否匿名（仅Pro有效）
	 */
	public void replyXml (String message, boolean isAnonymous) {
		reply (message, ChatRobotSendMessageType.xml, isAnonymous);
	}

	/**
	 * 回复Xml消息
	 *
	 * @param message 内容
	 */
	public void replyXml (String message) {
		reply (message, ChatRobotSendMessageType.xml, false);
	}

	/**
	 * 回复Xml消息
	 *
	 * @param message     内容
	 * @param isAnonymous 是否匿名（仅Pro有效）
	 */
	public void replyXml (Object message, boolean isAnonymous) {
		reply (message, ChatRobotSendMessageType.xml, isAnonymous);
	}

	/**
	 * 回复Xml消息
	 *
	 * @param message 内容
	 */
	public void replyXml (Object message) {
		reply (message, ChatRobotSendMessageType.xml, false);
	}

	/**
	 * 回复Xml消息
	 *
	 * @param format      格式化字符串
	 * @param isAnonymous 是否匿名（仅Pro有效）
	 * @param args        可变长参数
	 */
	public void replyXml (String format, boolean isAnonymous, Object... args) {
		reply (format, ChatRobotSendMessageType.xml, isAnonymous, args);
	}

	/**
	 * 回复Xml消息
	 *
	 * @param format 格式化字符串
	 * @param args   可变长参数
	 */
	public void replyXml (String format, Object... args) {
		reply (format, ChatRobotSendMessageType.xml, false, args);
	}

	/**
	 * 是否为语音消息
	 */
	public ChatRobotVoiceMessageResult isVoice () {
		return ChatRobotAPI.isVoiceMessage (text);
	}

	/**
	 * 消息中是否包含图片
	 *
	 * @return 是否包含, 提取出来的图片GUID
	 */
	public Pair<Boolean, List<String>> containsPicture () {
		return ChatRobotAPI.containsPictureInMessage (text);
	}

	/**
	 * 消息中是否包含艾特
	 *
	 * @return 是否包含, 被艾特的人
	 */
	public Pair<Boolean, List<Long>> containsAt () {
		return ChatRobotAPI.containsAtInMessage (text);
	}

	/**
	 * 是否为闪照消息
	 */
	public boolean isFlashPicture () {
		return ChatRobotAPI.isFlashPictureMessage (text);
	}

	/**
	 * 发送消息
	 *
	 * @param chatRobot   调用实例
	 * @param type        目标类型
	 * @param robot       机器人
	 * @param message     内容
	 * @param group       目标群
	 * @param qq          目标QQ
	 * @param isAnonymous 是否匿名（仅Pro有效）
	 * @param sendType    文本、Json、Xml消息
	 */
	public static void send (ChatRobot chatRobot, ChatRobotMessageType type, long robot, String message, long group, long qq,
							 boolean isAnonymous, ChatRobotSendMessageType sendType
	) {
		switch (type) {
			case friend:
				switch (sendType) {
					case text:
						chatRobot.sendFriendMessage (robot, qq, message);
						break;
					case json:
						chatRobot.sendFriendJsonMessage (robot, qq, message);
						break;
					case xml:
						chatRobot.sendFriendXmlMessage (robot, qq, message);
						break;
					default:
						new Exception (String.format ("%s.%s", type, sendType)).printStackTrace ();
						break;
				}
				break;
			case group:
				switch (sendType) {
					case text:
						chatRobot.sendGroupMessage (robot, group, message, isAnonymous);
						break;
					case json:
						chatRobot.sendGroupJsonMessage (robot, group, message, isAnonymous);
						break;
					case xml:
						chatRobot.sendGroupXmlMessage (robot, group, message, isAnonymous);
						break;
					default:
						new Exception (String.format ("%s.%s", type, sendType)).printStackTrace ();
						break;
				}
				break;
			default:
				new Exception (String.format ("%s.%s", type, sendType)).printStackTrace ();
				break;
		}
	}

	/**
	 * 发送消息
	 *
	 * @param chatRobot   调用实例
	 * @param type        目标类型
	 * @param robot       机器人
	 * @param message     内容
	 * @param group       目标群
	 * @param qq          目标QQ
	 * @param isAnonymous 是否匿名（仅Pro有效）
	 * @param sendType    文本、Json、Xml消息
	 */
	public static void send (ChatRobot chatRobot, ChatRobotMessageType type, long robot, Object message, long group, long qq,
							 boolean isAnonymous, ChatRobotSendMessageType sendType
	) {
		send (chatRobot, type, robot, message == null ? null : message.toString (), group, qq, isAnonymous, sendType);
	}

	/**
	 * 发送消息
	 *
	 * @param chatRobot   调用实例
	 * @param type        目标类型
	 * @param robot       机器人
	 * @param format      格式化字符串
	 * @param group       目标群
	 * @param qq          目标QQ
	 * @param isAnonymous 是否匿名（仅Pro有效）
	 * @param sendType    文本、Json、Xml消息
	 * @param args        可变长参数
	 */
	public static void send (ChatRobot chatRobot, ChatRobotMessageType type, long robot, String format, long group, long qq,
							 boolean isAnonymous, ChatRobotSendMessageType sendType, Object... args
	) {
		send (chatRobot, type, robot, String.format (format, args), group, qq, isAnonymous, sendType);
	}

	/**
	 * 3
	 * 返回消息内容（ChatRobotMessage.Text）
	 */
	@Override
	public String toString () {
		return text;
	}

	private void reply (String message, ChatRobotSendMessageType type, boolean isAnonymous) {
		send (chatRobot, this.type, robot, message, group, qq, isAnonymous, type);
	}

	private void reply (Object message, ChatRobotSendMessageType type, boolean isAnonymous) {
		send (chatRobot, this.type, robot, message == null ? null : message.toString (), group, qq, isAnonymous, type);
	}

	private void reply (String format, ChatRobotSendMessageType type, boolean isAnonymous, Object... args) {
		send (chatRobot, this.type, robot, String.format (format, args), group, qq, isAnonymous, type);
	}

}